### DISCUSSION TITLE
How to use `.eq_any(subquery)` for a parameter?

### DISCUSSION THREAD

**Discussion by hasezoey** [diesel-rs/diesel#3705] [Q&A]
I have been successfully using the answer from #3688, but have run into a new issue, that is when i use `.eq_any(subquery)` with the same typing as from #3688, then rust complains that `the trait "AppearsOnTable<query_source::joins::Join<tableA::table, tableC::table, Inner>>" is not implemented for "P"`.

I have already tried using `.eq_any(subquery.into_boxed())`, which is not `+ Clone` or `.eq_any(subquery).into_boxed()` where `.into_boxed()` is not implemented. Also tried using a `diesel::alias!`, which did not change the error.

Trying to add the suggested type is also not possible because `query_source::joins` is private (and i have not found a equivalent) `+ diesel::AppearsOnTable<diesel::query_source::joins::Join<tableA::table, tableC::table, diesel::query_source::joins::Inner>>` and i also dont think this is practical to be added if the conditions could also be multi-level.

using plain `.eq_any(vec![0,1,2,3])` does not throw any error

reproduction repository & branch: https://github.com/hasezoey/diesel-error-testing/tree/clone-as-parameter

full error:

```txt
error[E0277]: the trait bound `P: AppearsOnTable<query_source::joins::Join<tableA::table, tableC::table, Inner>>` is not satisfied
  --> src.rs:78:24
   |
78 |             delete_many_c(conn, tablec_filter)?;
   |             -------------       ^^^^^^^^^^^^^ the trait `AppearsOnTable<query_source::joins::Join<tableA::table, tableC::table, Inner>>` is not implemented for `P`
   |             |
   |             required by a bound introduced by this call
   |
   = note: required for `diesel::query_builder::where_clause::WhereClause<P>` to implement `diesel::query_builder::where_clause::ValidWhereClause<FromClause<query_source::joins::Join<tableA::table, tableC::table, Inner>>>`
   = note: required for `SelectStatement<FromClause<table>, SelectClause<rowid>, NoDistinctClause, ...>` to implement `expression::subselect::ValidSubselect<tableC::table>`
   = note: the full type name has been written to '/mnt/projects/rust/diesel-error-testing/target/debug/deps/mh-7e680b0187f653a7.long-type-4096700251850590849.txt'
   = note: required for `Subselect<SelectStatement<FromClause<table>, SelectClause<rowid>, ..., ...>, ...>` to implement `SelectableExpression<tableC::table>`
   = note: the full type name has been written to '/mnt/projects/rust/diesel-error-testing/target/debug/deps/mh-7e680b0187f653a7.long-type-15327216425325595087.txt'
   = note: 2 redundant requirements hidden
   = note: required for `Grouped<In<ref_tablea, Subselect<SelectStatement<..., ..., ..., ...>, ...>>>` to implement `SelectableExpression<tableC::table>`
   = note: the full type name has been written to '/mnt/projects/rust/diesel-error-testing/target/debug/deps/mh-7e680b0187f653a7.long-type-8685276965661447996.txt'
   = note: required for `Grouped<In<ref_tablea, Subselect<SelectStatement<..., ..., ..., ...>, ...>>>` to implement `BoxableExpression<tableC::table, Sqlite>`
   = note: the full type name has been written to '/mnt/projects/rust/diesel-error-testing/target/debug/deps/mh-7e680b0187f653a7.long-type-8685276965661447996.txt'
note: required by a bound in `delete_many_c`
  --> src.rs:93:5
   |
91 |   pub fn delete_many_c<P>(db: &mut SqliteConnection, predicate: P) -> QueryResult<()>
   |          ------------- required by a bound in this function
92 |   where
93 |       P: diesel::BoxableExpression<
   |  ________^
94 | |             tableC::table,
95 | |             diesel::sqlite::Sqlite,
96 | |             SqlType = diesel::sql_types::Bool,
97 | |         > + diesel::expression::NonAggregate
   | |_________^ required by this bound in `delete_many_c`
help: consider further restricting this bound
   |
64 |         + Clone + diesel::AppearsOnTable<diesel::query_source::joins::Join<tableA::table, tableC::table, diesel::query_source::joins::Inner>>
   |                 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
```

**Comment by hasezoey** [comment]
i had advanced the error to "`Clone` is not implemented for `P`" when calling `delete_many_c` and having `.eq_any(subquery.into_boxed())` called, so i resolved to remove the `Clone` constraint (though i would rather have it stay, if i need it) and subsequently resorted to use `ON DELETE CASCADE` in on the tables instead of manually cascading, though i would still like a answer on how to properly fix this (using `.eq_any(subquery)` as a parameter)

**Comment by weiznich** [comment]
Adding this bound to the `delete_many` function fixes the error:

```rust
    diesel::dsl::EqAny<
        tableC::ref_tablea,
        diesel::dsl::Select<diesel::dsl::Filter<tableA::table, P>, tableA::rowid>,
    >: diesel::BoxableExpression<
        tableC::table,
        diesel::sqlite::Sqlite,
        SqlType = diesel::sql_types::Bool,
    >,
```

It essentially just uplifts the bound from `delete_many_c` to the outer `delete_many` function by building the correct type. You can construct those types by just following the type (aliases) returned by `QueryDsl` and `ExpressionMethod` methods.

**Reply by hasezoey** [reply]
could you maybe explain why this is required explicitly for `EqAny` and what about if a different query is passed in that may be longer (more nested sub queries)? (the example given was minified to one level, but it may be multiple levels)

**Reply by weiznich** [reply]
It does not strictly require the `EqAny`, it just requires something in the trait prove chain to satisfy the `P: …` bound on the `delete_many_c`. I just went with that variant as that was relatively simple to write down. Technically speaking you could also go with much lower level bounds like as those suggested by the compiler. Practically speaking with diesel it's almost always easier to go with those kind of high level bounds as you then write usually only a single bound instead of a bunch of those low level bounds (if that's even possible due to private types).

If you do multiple different nested sub-queries you need to have bounds for all of them. That's just how rustc's trait bound checking is working and that's the reason why I generally advice people not to write these kind of generic code with diesel outside of foundational crates.

**Reply by hasezoey** [reply]
>  That's just how rustc's trait bound checking is working and that's the reason why I generally advice people not to write these kind of generic code with diesel outside of foundational crates.

somewhat sad to know, i had a more real-world example like:
(style `current_table -> referenced_table (reference_to)`)
- franchise
- media -> franchise (_id)
- media_entries -> media (_id)
- media_consumed -> media_entries (_id)

and when i wanted to delete a `franchise`, i would either need to go down and add a `delete_by_franchise` for every lower table (and increasingly add function like `delete_by_media`, etc) instead of just having a `delete_by(predicate)` where the previous one could mutate the input predicate to the next `delete_by` (as shown in the example code) and that same `delete_by` could also be used stand-alone, i though this would be possible because sql (at least in sqlite) it could be easily subqueried, it may not have been the most optimal query, but it would have been great for this case.
In the end for this case (as noted on another comment) i had gone with `ON DELETE CASCADE`, but i had initially wanted a little more control, without too much overhead (like extra function specific to that case) and it had also worked great, until i had to use `eq_any`.

thanks for your answering though.
